src/plugins/discover/public/application/main/utils/get_switch_index_pattern_app_state.ts

```diff
@@ -28,19 +28,17 @@ export function getSwitchIndexPatternAppState(
       )
     : currentColumns;
   const columns = nextColumns.length ? nextColumns : [];
-  // when switching from an index pattern with timeField to an index pattern without timeField
-  // filter out sorting by timeField in case it is set. index patterns without timeField don't
-  // prepend this field in the table, so in legacy grid you would need to add this column to
-  // remove sorting
-  const nextSort = getSortArray(currentSort, nextIndexPattern).filter((value) => {
-    return nextIndexPattern.timeFieldName || value[0] !== currentIndexPattern.timeFieldName;
-  });
+
+  let nextSort = getSortArray(currentSort, nextIndexPattern);
+  const [first, ...restSorting] = nextSort;
+  // replace primary timeFieldName if it was sorted
+  if (nextIndexPattern.timeFieldName && first && first[0] === currentIndexPattern.timeFieldName) {
+    nextSort = [[nextIndexPattern.timeFieldName, sortDirection], ...restSorting];
+  }
+
   return {
     index: nextIndexPattern.id,
     columns,
-    sort:
-      nextIndexPattern.timeFieldName && !nextSort.length
-        ? [[nextIndexPattern.timeFieldName, sortDirection]]
-        : nextSort,
+    sort: !nextSort.length ? [[nextIndexPattern.timeFieldName, sortDirection]] : nextSort,
```

Comment:
refactoring like this would, and switching to an index pattern without timefieldName, could create an sort entry like this `[,'desc']`, so the condition should stay `nextIndexPattern.timeFieldName && !nextSort.length`